### PR TITLE
[Merged by Bors] - feat: sheaves of modules generated by their sections

### PR DESCRIPTION
**Pull request by joelriou**
In this PR, we show the existence of colimits in categories of sheaves of modules, apply this to the particular case of coproducts of copies of the sheaf of rings in order to formalize the notion of generating family of (global) sections of a sheaf of modules.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/ModuleCat/Sheaf/Free.lean [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+noncomputable def free (I : Type u) : SheafOfModules.{u} R := ∐ (fun (_ : I) ↦ unit R)
+
+/-- The data of a morphism `free I ⟶ M` from a free sheaf of modules is
+equivalent to the data of a family `I → M.sections` of sections of `M`. -/
+noncomputable def freeHomEquiv (M : SheafOfModules.{u} R) {I : Type u} :
+    (free I ⟶ M) ≃ (I → M.sections) where
+  toFun f i := M.unitHomEquiv (Sigma.ι (fun (_ : I) ↦ unit R) i ≫ f)
+  invFun s := Sigma.desc (fun i ↦ M.unitHomEquiv.symm (s i))
+  left_inv s := Sigma.hom_ext _ _ (by simp)
+  right_inv f := by ext1 i; simp
```
Comment (erdOne) [inline Mathlib/Algebra/Category/ModuleCat/Sheaf/Free.lean:36]: Can we package this as an adjunction? (maybe TODO)

File: Mathlib/CategoryTheory/Sites/Over.lean [hunk 2]
```diff
@@ -199,4 +199,10 @@ abbrev overMapPullback (A : Type u') [Category.{v'} A] {X Y : C} (f : X ⟶ Y) :
 
 end GrothendieckTopology
 
+variable {J}
+
+/-- Given `F : Sheaf J A` and `X : C`, this is the pullback of `F` on `J.over X`. -/
+abbrev Sheaf.over {A : Type u'} [Category.{v'} A] (F : Sheaf J A) (X : C) :
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Sites/Over.lean:205]: Where did you use this?

Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/Over.lean:205]: Ah, not in this PR :-(, but it shall be used in #13464.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**